File: src/main/java/de/tectoast/emolga/utils/automation/structure/ModalConfigurator.java
package de.tectoast.emolga.utils.automation.structure;

import de.tectoast.jsolf.JSONObject;
import net.dv8tion.jda.api.EmbedBuilder;
import net.dv8tion.jda.api.entities.Member;
import net.dv8tion.jda.api.events.interaction.ModalInteractionEvent;
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent;
import net.dv8tion.jda.api.interactions.components.ActionRow;
import net.dv8tion.jda.api.interactions.components.Modal;
import net.dv8tion.jda.api.interactions.components.selections.SelectMenu;
import net.dv8tion.jda.api.interactions.components.text.TextInput;
import net.dv8tion.jda.api.interactions.modals.ModalMapping;

import java.awt.*;
import java.util.List;
import java.util.*;
import java.util.function.Function;
import java.util.stream.Collectors;

import static de.tectoast.emolga.commands.Command.getEmolgaJSON;
import static de.tectoast.emolga.commands.Command.saveEmolgaJSON;

public class ModalConfigurator {

    private final List<TextInput> actionRows = new LinkedList<>();
    private final Map<String, Function<String, Object>> mapper = new HashMap<>();
    private String id;
    private String title;


    private ModalConfigurator() {
    }

    public static ModalConfigurator create() {
        return new ModalConfigurator();
    }

    public Modal buildModal() {
        return buildModal(0);
    }

    public Modal buildModal(int page) {
        return Modal.create("modalconfigurator;" + id, title)
                .addActionRows(actionRows.stream().skip(page * 5L).limit(5).map(ActionRow::of).toList())
                .build();
    }

    public void handle(ModalInteractionEvent e) {
        List<ModalMapping> values = e.getValues();
        JSONObject o = getEmolgaJSON().createOrGetJSON("configuration").createOrGetJSON(e.getGuild().getIdLong()).createOrGetJSON(id);
        Member member = e.getMember();
        for (ModalMapping mm : values) {
            String id = mm.getId();
            String value = mm.getAsString();
            if (value.isBlank()) continue;
            Object mappedValue = mapper.containsKey(id) ? mapper.get(id).apply(value) : value;
            if (mappedValue == null) {
                e.replyEmbeds(new EmbedBuilder()
                        .setTitle("Das Argument %s ist ungültig für \"%s\"!".formatted(value, actionRows.stream()
                                .filter(ti -> ti.getId().equals(id)).findFirst().orElseThrow().getLabel()))
                        .setColor(0xFF0000)
                        .setFooter("Aufgerufen von %s (%s)".formatted(member.getEffectiveName(), member.getUser().getAsTag()))
                        .build()).queue();
                return;
            }
            o.put(id, mappedValue);
        }
        e.replyEmbeds(new EmbedBuilder()
                .setTitle("Deine Konfiguration wurde erfolgreich gespeichert!")
                .setColor(0x00FF00)
                .setFooter("Aufgerufen von %s (%s)".formatted(member.getEffectiveName(), member.getUser().getAsTag()))
                .build()).queue();
        saveEmolgaJSON();
    }

    public void initialize(SlashCommandInteractionEvent e) {
        if (actionRows.size() <= 5) {
            e.replyModal(buildModal()).queue();
            return;
        }
        EmbedBuilder embed = new EmbedBuilder()
                .setTitle("Welche Seite möchtest du sehen?")
                .setColor(Color.CYAN);
        SelectMenu.Builder sm = SelectMenu.create("modalconfigurator;" + id);
        for (int i = 0; i < (actionRows.size() / 5) + 1; i++) {
            int realSite = i + 1;
            embed.addField("Seite " + realSite,
                    actionRows.stream().skip(i * 5).limit(5)
                            .map(TextInput::getLabel).collect(Collectors.joining("\n")), false);
            sm.addOption("Seite " + realSite, String.valueOf(i));
        }
        e.replyEmbeds(embed.build()).addActionRow(sm.build()).queue();
    }

    public ModalConfigurator id(String id) {
        this.id = id;
        return this;
    }

    public ModalConfigurator title(String title) {
        this.title = title;
        return this;
    }

    public ModalConfigurator actionRows(TextInput... actionRows) {
        this.actionRows.addAll(Arrays.asList(actionRows));
        return this;
    }

    public ModalConfigurator mapper(Function<String, Object> mapper, String... ids) {
        for (String s : ids) {
            this.mapper.put(s, mapper);
        }
        return this;
    }

    @Override
    public String toString() {
        return "ModalConfigurator{" +
               "id='" + id + '\'' +
               ", title='" + title + '\'' +
               '}';
    }
}

File: src/main/kotlin/de/tectoast/emolga/utils/automation/structure/ModalConfigurator.kt
package de.tectoast.emolga.utils.automation.structure

import de.tectoast.emolga.commands.Command
import de.tectoast.emolga.utils.json.Configuration
import de.tectoast.emolga.utils.json.db
import net.dv8tion.jda.api.EmbedBuilder
import net.dv8tion.jda.api.events.interaction.ModalInteractionEvent
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent
import net.dv8tion.jda.api.interactions.components.ActionRow
import net.dv8tion.jda.api.interactions.components.selections.StringSelectMenu
import net.dv8tion.jda.api.interactions.components.text.TextInput
import net.dv8tion.jda.api.interactions.modals.Modal
import org.litote.kmongo.coroutine.updateOne
import org.litote.kmongo.eq
import java.awt.Color
import java.util.function.Function

class ModalConfigurator private constructor() {
    private val actionRows: MutableList<TextInput> = mutableListOf()
    private val mapper: MutableMap<String, Function<String, Any?>> = mutableMapOf()
    private var id: String? = null
    private var title: String? = null


    fun buildModal(page: Int = 0): Modal {
        return Modal.create("modalconfigurator;$id", title!!)
            .addComponents(
                actionRows.drop(page * 5).take(5).map { ActionRow.of(it) })
            .build()
    }

    suspend fun handle(e: ModalInteractionEvent) {
        val values = e.values
        val gid = e.guild!!.idLong
        val configuration = db.configuration.findOne(Configuration::guild eq gid) ?: Configuration(gid)
        val o = configuration.data.getOrPut(id!!) { mutableMapOf() }
        val member = e.member!!
        for (mm in values) {
            val id = mm.id
            val value = mm.asString
            if (value.isBlank()) continue
            val mappedValue = if (mapper.containsKey(id)) mapper[id]!!.apply(value) else value
            if (mappedValue == null) {
                e.replyEmbeds(
                    EmbedBuilder()
                        .setTitle(
                            "Das Argument $value ist ungültig für \"${actionRows.first { it.id == id }.label}\"!"
                        )
                        .setColor(0xFF0000)
                        .setFooter("Aufgerufen von ${member.effectiveName} (${member.user.effectiveName})")
                        .build()
                ).queue()
                return
            }
            o[id] = mappedValue.toString().toIntOrNull() ?: run {
                Command.sendToMe("$mappedValue in $id is not a long")
                return
            }
        }
        e.replyEmbeds(
            EmbedBuilder()
                .setTitle("Deine Konfiguration wurde erfolgreich gespeichert!")
                .setColor(0x00FF00)
                .setFooter("Aufgerufen von ${member.effectiveName} (${member.user.effectiveName}")
                .build()
        ).queue()
        db.configuration.updateOne(configuration)
    }

    fun initialize(e: SlashCommandInteractionEvent) {
        if (actionRows.size <= 5) {
            e.replyModal(buildModal()).queue()
            return
        }
        val embed = EmbedBuilder()
            .setTitle("Welche Seite möchtest du sehen?")
            .setColor(Color.CYAN)
        val sm = StringSelectMenu.create("modalconfigurator;$id")
        for (i in 0 until actionRows.size / 5 + 1) {
            val realSite = i + 1
            embed.addField(
                "Seite $realSite",
                actionRows.drop(i * 5).take(5)
                    .joinToString("\n") { obj: TextInput -> obj.label }, false
            )
            sm.addOption("Seite $realSite", i.toString())
        }
        e.replyEmbeds(embed.build()).addActionRow(sm.build()).queue()
    }

    fun id(id: String?): ModalConfigurator {
        this.id = id
        return this
    }

    fun title(title: String?): ModalConfigurator {
        this.title = title
        return this
    }

    fun actionRows(vararg actionRows: TextInput): ModalConfigurator {
        this.actionRows.addAll(listOf(*actionRows))
        return this
    }

    fun actionRows(actionRows: List<TextInput>): ModalConfigurator {
        this.actionRows.addAll(actionRows)
        return this
    }

    fun mapper(mapper: Function<String, Any?>, vararg ids: String): ModalConfigurator {
        for (s in ids) {
            this.mapper[s] = mapper
        }
        return this
    }

    override fun toString(): String {
        return "ModalConfigurator{" +
                "id='" + id + '\'' +
                ", title='" + title + '\'' +
                '}'
    }

    companion object {
        fun create(): ModalConfigurator {
            return ModalConfigurator()
        }
    }
}
